File: boj/problem2309/Main.java
package problem2309;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    private static final int N = 9;
    private static final int REAL_N = 7;
    private static ArrayList<int[]> answers = new ArrayList<>();

    public static void backtrack(int depth, boolean[] visited, int[] arr, int[] combs) {
        if (depth == REAL_N) {
            int[] temp = new int[N];
            int sum = 0;
            for (int i = 0; i < combs.length; i++) {
                sum += combs[i];
                temp[i] = combs[i];

            }

            if (sum == 100)
                answers.add(temp);

            return;
        }

        for (int i = depth; i < arr.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                combs[i] = arr[i];
                backtrack(depth + 1, visited, arr, combs);
                combs[i] = 0;
                visited[i] = false;
            }
        }
    }

    public static String solve(int[] arr) {
        boolean[] visited = new boolean[N];
        int[] combs = new int[N];

        backtrack(0, visited, arr, combs);

        int[] answer = answers.get(0);
        Arrays.sort(answer);
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < answer.length; i++) {
            if (i == 0 && answer[i] != 0)
                sb.append(Integer.toString(answer[i]));
            else if (answer[i] != 0)
                sb.append("\n").append(Integer.toString(answer[i]));
        }

        return sb.toString();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            arr[i] = Integer.parseInt(st.nextToken());
        }

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(solve(arr));
        bw.flush();
    }
}

File: boj/problem2309/main.py
from typing import List


N = 9
real_N = 7
answers = []


def backtrack(depth: int, visited: List[bool], arr: List[int], combs: List[int]):
    if depth == real_N:
        temp = [v for v in combs]

        if sum(temp) == 100:
            answers.append(temp)

        return

    for i in range(depth, len(arr)):
        if not visited[i]:
            visited[i] = True
            combs.append(arr[i])
            backtrack(depth + 1, visited, arr, combs)
            combs.pop()
            visited[i] = False


def solve(arr: List[int]):
    visited = [False] * N
    combs = []

    backtrack(0, visited, arr, combs)

    return "\n".join(map(str, sorted(answers[0])))


def main():
    arr = []
    for i in range(N):
        height = int(input())
        arr.append(height)

    print(solve(arr))


if __name__ == "__main__":
    main()
